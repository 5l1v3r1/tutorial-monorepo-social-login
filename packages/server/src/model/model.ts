import { domain, val } from "@plumier/core"
import { collection, model } from "@plumier/mongoose"
import reflect from "tinspector"


export type UserRole = "Admin" | "User"

export interface LoginUser {
    userId: string,
    role: UserRole
}

@domain()
export class DomainBase {
    constructor(
        @val.optional()
        public createdAt: Date = new Date(),
        @val.optional()
        public updatedAt: Date = new Date(),
        @val.optional()
        public deleted: boolean = false
    ) { }
}

@collection()
export class Todo extends DomainBase {
    constructor(
        @val.length({ max: 64 })
        public title: string,
        @val.optional()
        public completed?: boolean
    ) { super() }
}

export const TodoModel = model(Todo)

@collection()
export class SocialLogin extends DomainBase {
    constructor(
        public socialId: string,
        public provider: "Facebook" | "Google" | "Github"
    ) { super() }
}

export const SocialLoginModel = model(SocialLogin)

@collection()
export class User extends DomainBase {
    constructor(
        public name: string,
        @val.email()
        @val.unique()
        public email:string,
        @val.optional()
        public picture: string="",
        public userName: string,
        public password: string,
        public role: UserRole,
        @reflect.array(SocialLogin)
        public socialLogin: SocialLogin[]
    ) { super() }
}

export const UserModel = model(User)
