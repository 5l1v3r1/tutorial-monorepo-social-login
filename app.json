{
  "addons": [

  ],
  "buildpacks": [
    {
      "url": "heroku/nodejs"
    }
  ],
  "env": {
    "DEBUG": {
      "required": true
    },
    "FACEBOOK_CLIENT_ID": {
      "required": true
    },
    "FACEBOOK_SECRET": {
      "required": true
    },
    "GITHUB_CLIENT_ID": {
      "required": true
    },
    "GITHUB_SECRET": {
      "required": true
    },
    "GOOGLE_CLIENT_ID": {
      "required": true
    },
    "GOOGLE_SECRET": {
      "required": true
    },
    "JWT_SECRET": {
      "required": true
    },
    "MONGODB_URI": {
      "required": true
    },
    "PORT": {
      "required": true
    }
  },
  "formation": {
    "web": {
      "quantity": 1
    }
  },
  "name": "tutorial-monorepo-social-login",
  "scripts": {
  },
  "stack": "heroku-18"
}
