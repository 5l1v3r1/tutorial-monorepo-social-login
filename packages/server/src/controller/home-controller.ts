import "@plumier/serve-static"
import {route, response, authorize} from "plumier"
import { join } from "path"

export class HomeController {
    @authorize.public()
    @route.get("/")
    @route.get("/login")
    index(){
        return response.file(join(__dirname, "../../public/index.html"))
    }
}