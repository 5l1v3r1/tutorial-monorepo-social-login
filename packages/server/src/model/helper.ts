import { SchemaGenerator } from "@plumier/mongoose"
import mongoose from "mongoose"

export const schemaGenerator: SchemaGenerator = def => {
    if (def.deleted)
        def.deleted = { type: Boolean, default: false }
    const schema = new mongoose.Schema(def, { timestamps: true })
    schema.set("toJSON", { virtuals: true, versionKey:false })
    return schema
}