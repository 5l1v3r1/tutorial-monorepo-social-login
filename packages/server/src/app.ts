import { MongooseFacility } from "@plumier/mongoose"
import dotenv from "dotenv"
import Koa from "koa"
import { join } from "path"
import Plumier, { Configuration, WebApiFacility } from "plumier"

import { schemaGenerator } from "./model/helper"
import { JwtAuthFacility } from "@plumier/jwt"


dotenv.config({ path: join(__dirname, "../../../", ".env") })

export function createApp(config?: Partial<Configuration> & { mongoDbUri?: string }): Promise<Koa> {
    return new Plumier()
        .set(config || {})
        .set(new WebApiFacility())
        .set(new MongooseFacility({ uri: config && config.mongoDbUri || process.env.MONGODB_URI, schemaGenerator }))
        .set(new JwtAuthFacility({ secret: process.env.JWT_SECRET }))
        .initialize()
}
