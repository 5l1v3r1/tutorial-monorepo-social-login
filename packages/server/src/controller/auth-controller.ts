import {
    FacebookLoginStatus,
    FacebookProvider,
    GitHubLoginStatus,
    GitHubProvider,
    GoogleLoginStatus,
    GoogleProvider,
    oAuthCallback,
    SocialLoginStatus,
} from "@plumier/social-login"
import { sign } from "jsonwebtoken"
import { bind, response, route, authorize, HttpStatusError, val, ActionResult } from "plumier"
import bcrypt from "bcrypt"

import { LoginUser, SocialLogin, SocialLoginModel, User, UserModel, UserRole } from "../model/model"

type Provider = "Github" | "Facebook" | "Google"

function createAuthCookie(token: string) {
    return `Authorization=${token}; HttpOnly; SameSite=Strict; Path=/`
}

@authorize.public()
export class AuthController {

    // this endpoint used to exchange token with cookie
    // usually used by social login
    @route.post()
    async exchange(@bind.user() user: LoginUser) {
        const token = sign(<LoginUser>{ userId: user.userId, role: user.role }, process.env.JWT_SECRET)
        return new ActionResult({ accessToken: token })
            .setHeader("set-cookie", createAuthCookie(token))
    }

    @route.post()
    async login(@val.email() email: string, password: string) {
        const user = await UserModel.findOne({ email })
        if (user && await bcrypt.compare(password, user.password)) {
            const token = sign(<LoginUser>{ userId: user.id, role: user.role }, process.env.JWT_SECRET)
            return new ActionResult({ accessToken: token })
                .setHeader("set-cookie", createAuthCookie(token))
        }
        else throw new HttpStatusError(422, "Invalid username or password")
    }

    async logout() {
        return new ActionResult()
            .setHeader("set-cookie", createAuthCookie(""))
    }
}


@route.root("/auth")
@authorize.public()
export class SocialLoginController {
    @route.ignore()
    private async loginOrRegister<T>(login: SocialLoginStatus<T>, provider: Provider, transform: (data: T) => [string, Partial<User>]) {
        if (login.status === "Success") {
            const [socialId, user] = transform(login.data!)
            const socialLogin = await SocialLoginModel.findOne({ provider: provider, socialId: socialId })
            const savedUser = socialLogin && await UserModel.findOne({ socialLogin: socialLogin.id })
            let accessToken
            if (savedUser) {
                accessToken = sign(<LoginUser>{ userId: savedUser.id, role: savedUser.role }, process.env.JWT_SECRET)
            }
            else {
                const socialLogin = await new SocialLoginModel(<SocialLogin>{ socialId, provider }).save()
                const newUser = await new UserModel(<User>{ ...user!, role: "User", socialLogin: [socialLogin._id] }).save()
                accessToken = sign(<LoginUser>{ userId: newUser.id, role: newUser.role }, process.env.JWT_SECRET)
            }
            return response.callbackView({ status: login.status, accessToken })
                .setHeader("set-cookie", createAuthCookie(accessToken))
        }
        else
            return response.callbackView({ status: login.status })
    }

    @oAuthCallback(new FacebookProvider(process.env.FACEBOOK_CLIENT_ID, process.env.FACEBOOK_SECRET))
    async facebook(@bind.loginStatus() login: FacebookLoginStatus) {
        return this.loginOrRegister(login, "Facebook", x => {
            return [x.id, { name: x.name, picture: x.picture.data.url }]
        })
    }

    @oAuthCallback(new GoogleProvider(process.env.GOOGLE_CLIENT_ID, process.env.GOOGLE_SECRET))
    async google(@bind.loginStatus() login: GoogleLoginStatus) {
        return this.loginOrRegister(login, "Google", x => {
            return [x.id, { name: x.name, picture: x.picture }]
        })
    }

    @oAuthCallback(new GitHubProvider(process.env.GITHUB_CLIENT_ID, process.env.GITHUB_SECRET))
    async github(@bind.loginStatus() login: GitHubLoginStatus) {
        return this.loginOrRegister(login, "Github", x => {
            return [x.id.toString(), { name: x.name, picture: x.url }]
        })
    }
}