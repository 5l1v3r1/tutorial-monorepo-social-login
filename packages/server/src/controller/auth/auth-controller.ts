import {
    FacebookLoginStatus,
    FacebookProfile,
    FacebookProvider,
    GitHubLoginStatus,
    GitHubProfile,
    GitHubProvider,
    GoogleLoginStatus,
    GoogleProfile,
    GoogleProvider,
    oAuthCallback,
} from "@plumier/social-login"
import bcrypt from "bcrypt"
import Token from "csrf"
import { sign } from "jsonwebtoken"
import { Document } from "mongoose"
import { ActionResult, authorize, bind, HttpStatusError, response, route, val } from "plumier"

import { LoginUser, User, UserModel } from "../../model/model"

export function signToken(user: User & Document) {
    return sign(<LoginUser>{ userId: user.id, role: user.role }, process.env.JWT_SECRET)
}

//save global CSRF secret in memory
export const CSRF_SECRET = new Token().secretSync()

@route.root("/auth")
@authorize.public()
export class AuthController {

    //POST /auth/login
    @route.post()
    async login(@val.email() email: string, password: string) {
        const user = await UserModel.findOne({ email })
        if (user && await bcrypt.compare(password, user.password)) {
            const token = signToken(user)
            return new ActionResult({ accessToken: token })
                .setCookie("Authorization", token)
        }
        else throw new HttpStatusError(422, "Invalid username or password")
    }

    //GET /auth/logout
    async logout() {
        return new ActionResult()
            .setCookie("Authorization")
    }

    @route.ignore()
    private async loginOrRegister(status: "Success" | "Failed", state: string, user: Partial<User>) {
        //verify the CSRF token stored in state parameter (see SocialDialogController)
        if (!new Token().verify(CSRF_SECRET, state))
            return response.callbackView({ status: "Failed", message: "Invalid CSRF token" })
        //if login process (done by the callback) successful, proceed generate the JWT token
        if (status === "Success") {
            const savedUser = await UserModel.findOne({ provider: user.provider, socialId: user.socialId })
            let accessToken
            if (savedUser) {
                accessToken = signToken(savedUser)
            }
            else {
                const newUser = await new UserModel(<User>{ ...user!, role: "User" }).save()
                accessToken = signToken(newUser)
            }
            return response.callbackView({ status, accessToken })
                .setCookie("Authorization", accessToken)
        }
        else
            return response.callbackView({ status })
    }

    //GET /auth/facebook
    @oAuthCallback(new FacebookProvider(process.env.FACEBOOK_CLIENT_ID, process.env.FACEBOOK_SECRET))
    async facebook(@bind.loginStatus() login: FacebookLoginStatus, state: string) {
        const data = login.data || {} as FacebookProfile
        return this.loginOrRegister(login.status, state, { name: data.name, picture: data.picture.data.url, provider: "Facebook", socialId: data.id })
    }

    //GET /auth/google
    @oAuthCallback(new GoogleProvider(process.env.GOOGLE_CLIENT_ID, process.env.GOOGLE_SECRET))
    async google(@bind.loginStatus() login: GoogleLoginStatus, state: string) {
        const data = login.data || {} as GoogleProfile
        return this.loginOrRegister(login.status, state, { name: data.name, picture: data.picture, provider: "Google", socialId: data.id })
    }

    //GET /auth/github
    @oAuthCallback(new GitHubProvider(process.env.GITHUB_CLIENT_ID, process.env.GITHUB_SECRET))
    async github(@bind.loginStatus() login: GitHubLoginStatus, state: string) {
        const data = login.data || {} as GitHubProfile
        return this.loginOrRegister(login.status, state, { name: data.name, picture: data.url, provider: "Github", socialId: data.id.toString() })
    }
}