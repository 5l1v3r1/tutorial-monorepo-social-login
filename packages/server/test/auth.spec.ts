import { verify } from "jsonwebtoken"
import { ActionResult } from "plumier"

import { SocialLoginController } from "../src/controller/auth-controller"
import { LoginUser } from "../src/model/model"
import { appStub, AppStub } from "./helper/stub.app"
import stub from "./helper/stub"

function getLoginUserFromCallback(result: ActionResult) {
    const reg = result.body.match(/"accessToken":"(.*)"/)[1]
    return verify(reg, process.env.JWT_SECRET) as LoginUser
}

describe("Social Login", () => {
    let harness:AppStub
    beforeEach(async () => harness = await appStub())
    afterEach(async () => await harness.stop())


    it("Should able to login", async () => {
        const user = await stub.user.db({ provider: "Facebook", role: "User" })
        const controller = new SocialLoginController()
        const result = await controller.facebook(stub.facebook({ id: user.socialId }))
        const login = getLoginUserFromCallback(result)
        const savedUser = await stub.user.get({ provider: "Facebook", socialId: user.socialId })
        expect(savedUser!.toObject()).toMatchObject({ id: login.userId, role: login.role })
    })

    it("Should able to register and login", async () => {
        const controller = new SocialLoginController()
        const result = await controller.facebook(stub.facebook({ id: "12345678" }))
        const login = getLoginUserFromCallback(result)
        const user = await stub.user.get({ provider: "Facebook", socialId: "12345678" })
        expect(user!.toObject()).toMatchObject({ id: login.userId, role: login.role })
    })
})