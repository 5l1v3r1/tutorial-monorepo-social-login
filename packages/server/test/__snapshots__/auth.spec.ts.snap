// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Social Login Non registered user should able to login 1`] = `
"
    <!DOCTYPE html>
    <html>
        <title></title>
        <body>
            <div class=\\"container\\"></div>
            <script type=\\"text/javascript\\">
                var message = '{\\"status\\":\\"Success\\",}';
                (function(){
                    window.opener.onLogin(JSON.parse(message));
                })()
            </script>
        </body>
    </html>
    "
`;

exports[`Social Login Registered user should able to login 1`] = `
"
    <!DOCTYPE html>
    <html>
        <title></title>
        <body>
            <div class=\\"container\\"></div>
            <script type=\\"text/javascript\\">
                var message = '{\\"status\\":\\"Success\\",}';
                (function(){
                    window.opener.onLogin(JSON.parse(message));
                })()
            </script>
        </body>
    </html>
    "
`;
